CppExport/src/dependency_analysis/DependencyComposite.cpp

```diff
@@ -30,4 +30,22 @@ namespace CppExport {
 
 DependencyComposite::DependencyComposite(const QString& name) : name_(name) {}
 
+void DependencyComposite::addUnit(DependencyUnit* unit)
+{
+	units_.append(unit);
+	// assert every unit belongs to only one composite
+	Q_ASSERT(!unit->composite());
+	unit->setComposite(this);
+}
+
+QSet<const DependencyComposite*> DependencyComposite::dependencies(QList<DependencyUnit*>& allUnits)
+{
+	QSet<const DependencyComposite*> compositeDependencies;
+	for (DependencyUnit* unit : units_)
+		for (const DependencyUnit* unitDependency : unit->dependencies(allUnits))
+			compositeDependencies.insert(unitDependency->composite());
```

Comment:
It's possible (guaranteed almost always actually) that a unit will depend on another unit in the same composite. So you must check for self-reference here and ignore those cases. In fact you shouldn't ignore them, but you should make sure that within a Composite, all units are ordered topologically.